### PR TITLE
[Form][FrameworkBundle][Bridge] Add a DateInterval form type

### PR DESCRIPTION
**Pull request by MisatoTremor**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #13389 |
| License | MIT |
| Doc PR | symfony/symfony-docs#4817 |

Replaces #13390

### CODE REVIEW

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php [hunk 1]
```diff
@@ -0,0 +1,171 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
```
Comment (OskarStark) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php]: is the header copied from other files?

i think you should be named here @MisatoTremor

Comment (MisatoTremor) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php]: Yes, this is the standard copyright header.
I've put my name at the class author PHPDoc as usual. ;)

Comment (OskarStark) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php]: ok, i didn't know.

thank you!

**Comment by jewome62** [comment]
:cool:  !!!

When that will be integrated into the 2.8?

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToStringTransformer.php [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    public function reverseTransform($value)
+    {
+        if (empty($value)) {
+            return;
+        }
+        if (!is_string($value)) {
+            throw new TransformationFailedException('Expected a string.');
+        }
+        try {
+            if ($this->isISO8601($value)) {
```
Comment (stof) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToStringTransformer.php]: IMO, this if/else should be outside the try/catch, avoiding the need to catch TransformationFailedException as is (and it should use an early return)

Comment (MisatoTremor) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToStringTransformer.php]: Good catch, makes absolute sense to put it outside.
But is it really OK to just return null if the value is something the transformer does not expect?

Comment (MisatoTremor) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToStringTransformer.php]: @stof 
Just saw again that this throw > catch > throw pattern is also used in DateTimeToStringTransformer. Most probably I got it originally from there.

So maybe both should be changed (DateTimeToStringTransformer in another PR then) or just be left as is.

What do you think?

Comment (stof) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToStringTransformer.php]: Simplifying the DateTimeToStringTransformer in another PR makes sense yeah. but it should indeed be a separate PR

**Comment by MisatoTremor** [comment]
@stof Ok, simplified the transformer and also moved the format check out of the try/catch as this also was a throw > catch > throw.
Also fixed the deprecated form type by string name references and squashed commits.

AppVeyor build failure seems because of something else, as it also fails for the same tests in other PRs, etc. and Form component test passes.

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+/**
+ * Transforms between a normalized date interval and an interval string/array.
+ *
+ * @author Steffen Roßkamp <steffen.rosskamp@gimmickmedia.de>
+ */
+class DateIntervalToArrayTransformer implements DataTransformerInterface
+{
+    private $fields;
+    private $availableFields = array(
```
Comment (sstok) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php]: This can be made static as it actually never changes.

Same done in https://github.com/MisatoTremor/symfony/blob/add_dateinterval_type_2_8/src/Symfony/Component/Form/Extension/Core/DataTransformer/PercentToLocalizedStringTransformer.php#L29

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToStringTransformer.php [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    public function reverseTransform($value)
+    {
+        if (empty($value)) {
+            return;
+        }
+        if (!is_string($value)) {
+            throw new TransformationFailedException('Expected a string.');
+        }
+        if (!$this->isISO8601($value)) {
+            throw new TransformationFailedException('Non ISO 8601 date strings are not supported yet');
```
Comment (sstok) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToStringTransformer.php]: What about: `Only date intervals in the ISO 8601 format are supported.`

File: src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/DateIntervalTestCase.php [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Form\Tests\Extension\Core\DataTransformer;
+
+class DateIntervalTestCase extends \PHPUnit_Framework_TestCase
```
Comment (sstok) [inline src/Symfony/Component/Form/Tests/Extension/Core/DataTransformer/DateIntervalTestCase.php]: Should be marked abstract to prevent being executed as test-class.

**Comment by MisatoTremor** [comment]
Rebased on master
